utils/src/abi.rs

```diff
@@ -219,14 +227,21 @@ fn format_function_output_param(
 
 fn format_param(param: &Param, kind: String) -> String {
     // add `memory` if required (not needed for events, only for functions)
-    let is_memory = matches!(
+    let mut is_memory = matches!(
         param.kind,
-        ParamType::Array(_) |
-            ParamType::Bytes |
-            ParamType::String |
-            ParamType::FixedArray(_, _) |
-            ParamType::Tuple(_),
+        ParamType::Array(_)
+            | ParamType::Bytes
+            | ParamType::String
+            | ParamType::FixedArray(_, _)
+            | ParamType::Tuple(_),
     );
+    is_memory =
+        if (matches!(param.kind, ParamType::Tuple(_)) && param.internal_type.as_ref() == None) {
+            false
+        } else {
+            is_memory
+        };
```

Comment:
this can be merged into a single match statement:

```rust
match param.kind {
   ParamType::Array(_)
            | ParamType::Bytes
            | ParamType::String
            | ParamType::FixedArray(_, _) => true,
ParamType::Tuple(_) => {...}
_ => false
}
```

Reply:
changed it